File: server/workspace-manager.ts
import { mkdir, readFile, rename, rm, writeFile } from 'node:fs/promises';
import { randomUUID } from 'node:crypto';
import { basename, join, resolve } from 'node:path';
import { initialState } from '../src/demo.js';
import type { AppState, CreateWorkspaceInput, WorkspaceBootstrap, WorkspaceSummary } from '../src/types.js';
import { CareerService, dispatch as dispatchService } from './service.js';

interface CatalogEntry {
  id: string; name: string; relativeDirectory: string; createdAt: string; updatedAt: string; lastOpenedAt: string;
  archivedAt: string; deletedAt: string;
}
interface Catalog { version: 1; currentWorkspaceId: string; epoch: number; workspaces: CatalogEntry[] }

const now = () => new Date().toISOString();
const cleanName = (value: unknown) => {
  const name = typeof value === 'string' ? value.trim() : '';
  if (!name || [...name].length > 80) throw new Error('Workspace name must be between 1 and 80 characters.');
  return name;
};

export class PreviewWorkspaceManager {
  private catalog!: Catalog;
  private current!: CatalogEntry;
  currentService!: CareerService;
  private readonly catalogPath: string;
  constructor(readonly root: string, private readonly rootService: CareerService) { this.catalogPath = join(root, 'catalog.json'); }

  async init() {
    await mkdir(this.root, { recursive: true, mode: 0o700 });
    try {
      const value = JSON.parse(await readFile(this.catalogPath, 'utf8')) as Catalog;
      if (value.version !== 1 || !Array.isArray(value.workspaces) || typeof value.currentWorkspaceId !== 'string') throw new Error();
      this.catalog = value;
      this.current = this.entry(value.currentWorkspaceId);
      if (this.current.relativeDirectory === '.') this.currentService = this.rootService;
      else { await this.rootService.shutdown(); this.currentService = await this.rootService.createSibling(this.directory(this.current)).init(); }
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== 'ENOENT') throw new Error('Could not read the workspace catalog. Its existing file has been preserved.');
      const stamp = now();
      const entry: CatalogEntry = { id: randomUUID(), name: this.rootService.state.demo ? 'Sample workspace' : 'My workspace', relativeDirectory: '.', createdAt: stamp, updatedAt: stamp, lastOpenedAt: stamp, archivedAt: '', deletedAt: '' };
      this.catalog = { version: 1, currentWorkspaceId: entry.id, epoch: 1, workspaces: [entry] };
      this.current = entry; this.currentService = this.rootService;
      await this.persistCatalog();
    }
    return this;
  }

  private entry(id: string) {
    const item = this.catalog.workspaces.find(workspace => workspace.id === id);
    if (!item) throw new Error('Workspace not found.');
    return item;
  }
  private directory(entry: CatalogEntry) {
    if (entry.relativeDirectory === '.') return this.root;
    if (entry.relativeDirectory !== `workspaces/${entry.id}` || basename(entry.id) !== entry.id) throw new Error('Workspace path is invalid.');
    const directory = resolve(this.root, entry.relativeDirectory);
    if (!directory.startsWith(resolve(this.root) + '/')) throw new Error('Workspace path escapes the data directory.');
    return directory;
  }
  private async persistCatalog() {
    const temp = `${this.catalogPath}.tmp`;
    await writeFile(temp, JSON.stringify(this.catalog, null, 2), { mode: 0o600 });
    await rename(temp, this.catalogPath);
  }
  private ensureUnique(name: string, except = '') {
    if (this.catalog.workspaces.some(item => item.id !== except && item.name.localeCompare(name, undefined, { sensitivity: 'accent' }) === 0)) throw new Error('A workspace with that name already exists.');
  }
  private summary(entry: CatalogEntry, state?: AppState, unavailable = false): WorkspaceSummary {
    return { id: entry.id, name: entry.name, current: entry.id === this.current.id, demo: Boolean(state?.demo), opportunityCount: state?.roles.length || 0, savedCount: state?.saved.length || 0, profileComplete: Boolean(state?.profile.name.trim() && state?.profile.background.trim()), researchRunning: entry.id === this.current.id && this.currentService.isResearchActive(), createdAt: entry.createdAt, updatedAt: entry.updatedAt, lastOpenedAt: entry.lastOpenedAt, archivedAt: entry.archivedAt || undefined, deletedAt: entry.deletedAt || undefined, unavailable };
  }
  private async summaries() {
    const result: WorkspaceSummary[] = [];
    for (const entry of [...this.catalog.workspaces].sort((a, b) => b.lastOpenedAt.localeCompare(a.lastOpenedAt))) {
      if (entry.id === this.current.id) { result.push(this.summary(entry, this.currentService.getState())); continue; }
      try {
        const loaded = JSON.parse(await readFile(join(this.directory(entry), 'state.json'), 'utf8')) as AppState;
        result.push(this.summary(entry, loaded));
      } catch { result.push(this.summary(entry, undefined, true)); }
    }
    return result;
  }
  async bootstrap(): Promise<WorkspaceBootstrap> {
    const workspaces = await this.summaries();
    return { workspace: workspaces.find(item => item.id === this.current.id)!, workspaces, state: this.currentService.getState(), epoch: this.catalog.epoch };
  }
  private assertIdle() { if (this.currentService.isResearchActive()) throw new Error('Stop research before creating or switching workspaces.'); }
  async create(input: CreateWorkspaceInput) {
    this.assertIdle(); const name = cleanName(input?.name); this.ensureUnique(name);
    if (!['blank', 'copy', 'import'].includes(input?.mode)) throw new Error('Choose how to start the workspace.');
    const id = randomUUID(); const stamp = now();
    const entry: CatalogEntry = { id, name, relativeDirectory: `workspaces/${id}`, createdAt: stamp, updatedAt: stamp, lastOpenedAt: stamp, archivedAt: '', deletedAt: '' };
    const service = await this.rootService.createSibling(this.directory(entry)).init();
    try {
      if (input.mode === 'blank') await service.importState(initialState(false));
      if (input.mode === 'copy') {
        const copied = initialState(false); const source = this.currentService.getState();
        copied.profile = structuredClone(source.profile); copied.profileRevision = 1;
        copied.settings.jevEnabled = source.settings.jevEnabled; copied.settings.jevModel = source.settings.jevModel;
        await service.importState(copied);
      }
      if (input.mode === 'import') {
        if (!input.state) throw new Error('Choose a valid Waypoint workspace export.');
        await service.importState(input.state);
      }
      this.catalog.workspaces.push(entry); this.catalog.currentWorkspaceId = id; this.catalog.epoch++;
      await this.persistCatalog();
    } catch (error) { await service.shutdown(); await rm(this.directory(entry), { recursive: true, force: true }); throw error; }
    const old = this.currentService; this.current = entry; this.currentService = service; await old.shutdown();
    return this.bootstrap();
  }
  async switch(id: string) {
    this.assertIdle(); if (id === this.current.id) return this.bootstrap();
    const entry = this.entry(id); if (entry.archivedAt || entry.deletedAt) throw new Error('Restore this workspace before opening it.');
    let service: CareerService;
    try { service = await this.rootService.createSibling(this.directory(entry)).init(); } catch { throw new Error('The workspace could not be opened; the current workspace is unchanged.'); }
    const old = this.currentService; const stamp = now(); entry.lastOpenedAt = stamp; entry.updatedAt = stamp;
    this.catalog.currentWorkspaceId = id; this.catalog.epoch++;
    try { await this.persistCatalog(); } catch (error) { await service.shutdown(); throw error; }
    this.current = entry; this.currentService = service; await old.shutdown(); return this.bootstrap();
  }
  async renameWorkspace(id: string, requested: string) {
    const name = cleanName(requested); this.ensureUnique(name, id); const entry = this.entry(id);
    entry.name = name; entry.updatedAt = now(); await this.persistCatalog(); return this.bootstrap();
  }
  async lifecycle(id: string, action: 'archive' | 'restore' | 'delete') {
    const entry = this.entry(id);
    if (action !== 'restore' && id === this.current.id) throw new Error(`Switch workspaces before ${action === 'archive' ? 'archiving' : 'deleting'} the current one.`);
    if (action === 'archive') entry.archivedAt = now();
    if (action === 'delete') { const active = this.catalog.workspaces.filter(item => !item.archivedAt && !item.deletedAt).length; if (!entry.archivedAt && active <= 1) throw new Error('Keep at least one active workspace.'); entry.deletedAt = now(); }
    if (action === 'restore') { entry.archivedAt = ''; entry.deletedAt = ''; }
    entry.updatedAt = now(); await this.persistCatalog(); return this.bootstrap();
  }
  async permanentlyDelete(id: string, confirmation: string) {
    const entry = this.entry(id); if (id === this.current.id) throw new Error('The current workspace cannot be deleted.');
    if (!entry.deletedAt || confirmation !== entry.name) throw new Error('Type the workspace name exactly to delete it permanently.');
    if (entry.relativeDirectory !== '.') await rm(this.directory(entry), { recursive: true, force: true });
    else { await rm(join(this.root, 'state.json'), { force: true }); await rm(join(this.root, 'state.tmp'), { force: true }); }
    this.catalog.workspaces = this.catalog.workspaces.filter(item => item.id !== id); await this.persistCatalog(); return this.bootstrap();
  }
  async exportWorkspace(id: string) {
    if (id === this.current.id) return this.currentService.getState();
    return JSON.parse(await readFile(join(this.directory(this.entry(id)), 'state.json'), 'utf8')) as AppState;
  }
  async dispatch(method: string, args: unknown[]) {
    if (method === 'bootstrap') return this.bootstrap();
    if (method === 'createWorkspace') return this.create(args[0] as CreateWorkspaceInput);
    if (method === 'switchWorkspace') return this.switch(String(args[0]));
    if (method === 'renameWorkspace') return this.renameWorkspace(String(args[0]), String(args[1]));
    if (method === 'archiveWorkspace') return this.lifecycle(String(args[0]), 'archive');
    if (method === 'restoreWorkspace') return this.lifecycle(String(args[0]), 'restore');
    if (method === 'deleteWorkspace') return this.lifecycle(String(args[0]), 'delete');
    if (method === 'deleteWorkspacePermanently') return this.permanentlyDelete(String(args[0]), String(args[1]));
    if (method === 'exportWorkspace') return this.exportWorkspace(String(args[0]));
    return dispatchService(this.currentService, method, args);
  }
  async shutdown() { await this.currentService.shutdown(); }
}

File: internal/waypoint/workspace_manager.go
package waypoint

import (
	"errors"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
)

type WorkspaceManager struct {
	mu                sync.RWMutex
	root              string
	catalog           *CatalogRepository
	current           catalogEntry
	service           *Service
	epoch             uint64
	newService        func(string) *Service
	backgroundID      string
	backgroundService *Service
	backgroundOwned   bool
}

func NewWorkspaceManager(root string) *WorkspaceManager {
	return &WorkspaceManager{root: root, catalog: NewCatalogRepository(root), newService: NewService}
}

func (m *WorkspaceManager) Init() error {
	if err := m.catalog.Init(); err != nil {
		return err
	}
	currentID, found, err := m.catalog.CurrentID()
	if err != nil {
		_ = m.catalog.Close()
		return err
	}
	if !found {
		service := m.newService(m.root)
		if err := service.Init(); err != nil {
			_ = m.catalog.Close()
			return err
		}
		id, hasID, err := service.WorkspaceID()
		if err != nil {
			service.Shutdown()
			_ = m.catalog.Close()
			return err
		}
		if !hasID {
			id = uuid.NewString()
			if err := service.SetWorkspaceID(id); err != nil {
				service.Shutdown()
				_ = m.catalog.Close()
				return err
			}
		}
		now := nowISO()
		name := "My workspace"
		if service.State().Demo {
			name = "Sample workspace"
		}
		entry := catalogEntry{ID: id, Name: name, RelativeDirectory: ".", CreatedAt: now, UpdatedAt: now, LastOpenedAt: now}
		if err := m.catalog.Add(entry, true); err != nil {
			service.Shutdown()
			_ = m.catalog.Close()
			return err
		}
		m.current, m.service, m.epoch = entry, service, 1
		return nil
	}
	entry, err := m.catalog.Entry(currentID)
	if err != nil {
		_ = m.catalog.Close()
		return err
	}
	service, err := m.openService(entry)
	if err != nil {
		_ = m.catalog.Close()
		return errors.New("the selected workspace could not be opened; its files were preserved")
	}
	m.current, m.service, m.epoch = entry, service, 1
	return nil
}

func (m *WorkspaceManager) Close() {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.service != nil {
		m.service.Shutdown()
		m.service = nil
	}
	if m.backgroundOwned && m.backgroundService != nil {
		m.backgroundService.Shutdown()
	}
	m.backgroundService = nil
	_ = m.catalog.Close()
}

func (m *WorkspaceManager) Dispatch(method string, args []any) (any, error) {
	switch method {
	case "bootstrap":
		m.mu.RLock()
		defer m.mu.RUnlock()
		return m.bootstrapLocked()
	case "createWorkspace":
		var input CreateWorkspaceInput
		if err := decodeArg(args, 0, &input); err != nil {
			return nil, err
		}
		return m.Create(input)
	case "switchWorkspace":
		var id string
		if err := decodeArg(args, 0, &id); err != nil {
			return nil, err
		}
		return m.Switch(id)
	case "renameWorkspace":
		var id, name string
		if err := decodeArg(args, 0, &id); err != nil {
			return nil, err
		}
		if err := decodeArg(args, 1, &name); err != nil {
			return nil, err
		}
		return m.Rename(id, name)
	case "archiveWorkspace":
		return m.lifecycleArg(args, "archive")
	case "restoreWorkspace":
		return m.lifecycleArg(args, "restore")
	case "deleteWorkspace":
		return m.lifecycleArg(args, "delete")
	case "deleteWorkspacePermanently":
		var id, confirmation string
		if err := decodeArg(args, 0, &id); err != nil {
			return nil, err
		}
		if err := decodeArg(args, 1, &confirmation); err != nil {
			return nil, err
		}
		return m.DeletePermanently(id, confirmation)
	case "exportWorkspace":
		var id string
		if err := decodeArg(args, 0, &id); err != nil {
			return nil, err
		}
		return m.Export(id)
	default:
		m.mu.RLock()
		defer m.mu.RUnlock()
		if m.service == nil {
			return nil, errors.New("workspace is not available")
		}
		if method == "startResearch" && m.backgroundService != nil && m.backgroundService != m.service {
			return nil, errors.New("wait for research in the other workspace to finish")
		}
		return m.service.Dispatch(method, args)
	}
}

func (m *WorkspaceManager) lifecycleArg(args []any, action string) (WorkspaceBootstrap, error) {
	var id string
	if err := decodeArg(args, 0, &id); err != nil {
		return WorkspaceBootstrap{}, err
	}
	switch action {
	case "archive":
		return m.Archive(id)
	case "restore":
		return m.Restore(id)
	default:
		return m.MoveToDeleted(id)
	}
}

func (m *WorkspaceManager) Bootstrap() (WorkspaceBootstrap, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.bootstrapLocked()
}

func (m *WorkspaceManager) bootstrapLocked() (WorkspaceBootstrap, error) {
	items, err := m.summariesLocked()
	if err != nil {
		return WorkspaceBootstrap{}, err
	}
	var current WorkspaceSummary
	for _, item := range items {
		if item.ID == m.current.ID {
			current = item
			break
		}
	}
	return WorkspaceBootstrap{Workspace: current, Workspaces: items, State: m.service.State(), Epoch: m.epoch}, nil
}

func (m *WorkspaceManager) Create(input CreateWorkspaceInput) (WorkspaceBootstrap, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.service.ActiveResearch() || m.backgroundService != nil {
		return WorkspaceBootstrap{}, errors.New("stop research before creating or switching workspaces")
	}
	name, err := validateWorkspaceName(input.Name)
	if err != nil {
		return WorkspaceBootstrap{}, err
	}
	if input.Mode != "blank" && input.Mode != "copy" && input.Mode != "import" {
		return WorkspaceBootstrap{}, errors.New("choose how to start the workspace")
	}
	id := uuid.NewString()
	relative := filepath.Join("workspaces", id)
	entry := catalogEntry{ID: id, Name: name, RelativeDirectory: relative, CreatedAt: nowISO(), UpdatedAt: nowISO(), LastOpenedAt: nowISO()}
	directory := filepath.Join(m.root, relative)
	service := m.newService(directory)
	if err := service.Init(); err != nil {
		return WorkspaceBootstrap{}, err
	}
	ready := false
	defer func() {
		if !ready {
			service.Shutdown()
			for _, suffix := range []string{"", "-wal", "-shm"} {
				_ = os.Remove(filepath.Join(directory, databaseFilename) + suffix)
			}
			_ = os.Remove(directory)
		}
	}()
	switch input.Mode {
	case "blank":
		if _, err := service.Reset(); err != nil {
			return WorkspaceBootstrap{}, err
		}
	case "copy":
		state := InitialState(false)
		current := m.service.State()
		state.Profile = current.Profile
		state.ProfileRevision = 1
		state.Settings.JevEnabled = current.Settings.JevEnabled
		state.Settings.JevModel = current.Settings.JevModel
		if _, err := service.ImportState(state); err != nil {
			return WorkspaceBootstrap{}, err
		}
	case "import":
		if input.State == nil {
			return WorkspaceBootstrap{}, errors.New("choose a valid Waypoint workspace export")
		}
		if _, err := service.ImportState(*input.State); err != nil {
			return WorkspaceBootstrap{}, err
		}
	}
	if err := service.SetWorkspaceID(id); err != nil {
		return WorkspaceBootstrap{}, err
	}
	if err := m.catalog.Add(entry, true); err != nil {
		return WorkspaceBootstrap{}, err
	}
	old := m.service
	m.current, m.service = entry, service
	m.epoch++
	ready = true
	old.Shutdown()
	return m.bootstrapLocked()
}

func (m *WorkspaceManager) Switch(id string) (WorkspaceBootstrap, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if id == m.current.ID {
		return m.bootstrapLocked()
	}
	if m.service.ActiveResearch() || m.backgroundService != nil {
		return WorkspaceBootstrap{}, errors.New("stop research before switching workspaces")
	}
	entry, err := m.catalog.Entry(id)
	if err != nil {
		return WorkspaceBootstrap{}, err
	}
	if entry.ArchivedAt != "" || entry.DeletedAt != "" {
		return WorkspaceBootstrap{}, errors.New("restore this workspace before opening it")
	}
	service, err := m.openService(entry)
	if err != nil {
		return WorkspaceBootstrap{}, errors.New("the workspace could not be opened; the current workspace is unchanged")
	}
	if err := m.catalog.SetCurrent(id); err != nil {
		service.Shutdown()
		return WorkspaceBootstrap{}, err
	}
	entry, _ = m.catalog.Entry(id)
	old := m.service
	m.current, m.service = entry, service
	m.epoch++
	old.Shutdown()
	return m.bootstrapLocked()
}

func (m *WorkspaceManager) Rename(id, requested string) (WorkspaceBootstrap, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	name, err := validateWorkspaceName(requested)
	if err != nil {
		return WorkspaceBootstrap{}, err
	}
	if err := m.catalog.Rename(id, name); err != nil {
		return WorkspaceBootstrap{}, err
	}
	if id == m.current.ID {
		m.current.Name = name
	}
	return m.bootstrapLocked()
}

func (m *WorkspaceManager) Archive(id string) (WorkspaceBootstrap, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if id == m.backgroundID {
		return WorkspaceBootstrap{}, errors.New("wait for scheduled research to finish before archiving this workspace")
	}
	if id == m.current.ID {
		return WorkspaceBootstrap{}, errors.New("switch workspaces before archiving the current one")
	}
	if err := m.catalog.SetLifecycle(id, "archived_at", nowISO()); err != nil {
		return WorkspaceBootstrap{}, err
	}
	return m.bootstrapLocked()
}

func (m *WorkspaceManager) Restore(id string) (WorkspaceBootstrap, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if err := m.catalog.Restore(id); err != nil {
		return WorkspaceBootstrap{}, err
	}
	return m.bootstrapLocked()
}

func (m *WorkspaceManager) MoveToDeleted(id string) (WorkspaceBootstrap, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if id == m.backgroundID {
		return WorkspaceBootstrap{}, errors.New("wait for scheduled research to finish before deleting this workspace")
	}
	if id == m.current.ID {
		return WorkspaceBootstrap{}, errors.New("switch workspaces before deleting the current one")
	}
	entry, err := m.catalog.Entry(id)
	if err != nil {
		return WorkspaceBootstrap{}, err
	}
	count, err := m.catalog.ActiveCount()
	if err != nil {
		return WorkspaceBootstrap{}, err
	}
	if entry.ArchivedAt == "" && count <= 1 {
		return WorkspaceBootstrap{}, errors.New("keep at least one active workspace")
	}
	if err := m.catalog.SetLifecycle(id, "deleted_at", nowISO()); err != nil {
		return WorkspaceBootstrap{}, err
	}
	return m.bootstrapLocked()
}

func (m *WorkspaceManager) DeletePermanently(id, confirmation string) (WorkspaceBootstrap, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if id == m.backgroundID {
		return WorkspaceBootstrap{}, errors.New("wait for scheduled research to finish before deleting this workspace")
	}
	if id == m.current.ID {
		return WorkspaceBootstrap{}, errors.New("the current workspace cannot be deleted")
	}
	entry, err := m.catalog.Entry(id)
	if err != nil {
		return WorkspaceBootstrap{}, err
	}
	if entry.DeletedAt == "" || confirmation != entry.Name {
		return WorkspaceBootstrap{}, errors.New("type the workspace name exactly to delete it permanently")
	}
	directory, err := m.directoryFor(entry)
	if err != nil {
		return WorkspaceBootstrap{}, err
	}
	if entry.RelativeDirectory != "." {
		items, readErr := os.ReadDir(directory)
		if readErr != nil && !errors.Is(readErr, os.ErrNotExist) {
			return WorkspaceBootstrap{}, readErr
		}
		allowed := map[string]bool{databaseFilename: true, databaseFilename + "-wal": true, databaseFilename + "-shm": true}
		for _, item := range items {
			if !allowed[item.Name()] {
				return WorkspaceBootstrap{}, errors.New("the workspace directory contains unexpected files; nothing was deleted")
			}
		}
	}
	for _, suffix := range []string{"", "-wal", "-shm"} {
		if err := os.Remove(filepath.Join(directory, databaseFilename) + suffix); err != nil && !errors.Is(err, os.ErrNotExist) {
			return WorkspaceBootstrap{}, err
		}
	}
	if entry.RelativeDirectory != "." {
		if err := os.Remove(directory); err != nil && !errors.Is(err, os.ErrNotExist) {
			return WorkspaceBootstrap{}, err
		}
	}
	if err := m.catalog.Remove(id); err != nil {
		return WorkspaceBootstrap{}, err
	}
	return m.bootstrapLocked()
}

func (m *WorkspaceManager) Export(id string) (State, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	if id == m.current.ID {
		return m.service.State(), nil
	}
	entry, err := m.catalog.Entry(id)
	if err != nil {
		return State{}, err
	}
	service, err := m.openService(entry)
	if err != nil {
		return State{}, err
	}
	defer service.Shutdown()
	return service.State(), nil
}

func (m *WorkspaceManager) CurrentService() *Service {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.service
}

type ScheduledWorkspaceRun struct {
	Service       *Service
	WorkspaceID   string
	WorkspaceName string
	Before        map[string]bool
}

// StartNextScheduledResearch starts at most one due run across active workspaces.
// The caller must call FinishScheduledResearch after the returned run settles.
func (m *WorkspaceManager) StartNextScheduledResearch(now time.Time) (ScheduledWorkspaceRun, bool, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.backgroundService != nil || m.service == nil || m.service.ActiveResearch() {
		return ScheduledWorkspaceRun{}, false, nil
	}
	entries, err := m.catalog.Entries()
	if err != nil {
		return ScheduledWorkspaceRun{}, false, err
	}
	for _, entry := range entries {
		if entry.ArchivedAt != "" || entry.DeletedAt != "" {
			continue
		}
		service, owned := m.service, false
		if entry.ID != m.current.ID {
			service, err = m.openService(entry)
			if err != nil {
				continue
			}
			owned = true
		}
		before := map[string]bool{}
		for _, role := range service.State().Roles {
			before[role.ID] = true
		}
		started, startErr := service.StartScheduledResearch(now)
		if startErr != nil || !started {
			if owned {
				service.Shutdown()
			}
			if startErr != nil {
				return ScheduledWorkspaceRun{}, false, startErr
			}
			continue
		}
		m.backgroundID, m.backgroundService, m.backgroundOwned = entry.ID, service, owned
		return ScheduledWorkspaceRun{Service: service, WorkspaceID: entry.ID, WorkspaceName: entry.Name, Before: before}, true, nil
	}
	return ScheduledWorkspaceRun{}, false, nil
}

func (m *WorkspaceManager) FinishScheduledResearch(service *Service) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.backgroundService != service {
		return
	}
	if m.backgroundOwned {
		service.Shutdown()
	}
	m.backgroundID, m.backgroundService, m.backgroundOwned = "", nil, false
}

func (m *WorkspaceManager) summariesLocked() ([]WorkspaceSummary, error) {
	entries, err := m.catalog.Entries()
	if err != nil {
		return nil, err
	}
	items := make([]WorkspaceSummary, 0, len(entries))
	for _, entry := range entries {
		if entry.ID == m.current.ID {
			items = append(items, summaryFromState(entry, m.service.State(), true, m.service.ActiveResearch()))
			continue
		}
		if entry.ID == m.backgroundID && m.backgroundService != nil {
			items = append(items, summaryFromState(entry, m.backgroundService.State(), false, true))
			continue
		}
		directory, pathErr := m.directoryFor(entry)
		if pathErr != nil {
			items = append(items, unavailableSummary(entry))
			continue
		}
		if _, statErr := os.Stat(databasePath(directory)); statErr != nil {
			items = append(items, unavailableSummary(entry))
			continue
		}
		repository := NewSQLiteRepository(directory)
		if err := repository.Init(); err != nil {
			items = append(items, unavailableSummary(entry))
			continue
		}
		state, found, loadErr := repository.Load()
		_ = repository.Close()
		if loadErr != nil || !found {
			items = append(items, unavailableSummary(entry))
			continue
		}
		items = append(items, summaryFromState(entry, state, false, false))
	}
	return items, nil
}

func summaryFromState(entry catalogEntry, state State, current, running bool) WorkspaceSummary {
	return WorkspaceSummary{
		ID: entry.ID, Name: entry.Name, Current: current, Demo: state.Demo,
		OpportunityCount: len(state.Roles), SavedCount: len(state.Saved),
		ProfileComplete: strings.TrimSpace(state.Profile.Name) != "" && strings.TrimSpace(state.Profile.Background) != "",
		ResearchRunning: running, CreatedAt: entry.CreatedAt, UpdatedAt: entry.UpdatedAt,
		LastOpenedAt: entry.LastOpenedAt, ArchivedAt: entry.ArchivedAt, DeletedAt: entry.DeletedAt,
	}
}

func unavailableSummary(entry catalogEntry) WorkspaceSummary {
	return WorkspaceSummary{ID: entry.ID, Name: entry.Name, CreatedAt: entry.CreatedAt, UpdatedAt: entry.UpdatedAt, LastOpenedAt: entry.LastOpenedAt, ArchivedAt: entry.ArchivedAt, DeletedAt: entry.DeletedAt, Unavailable: true}
}

func (m *WorkspaceManager) openService(entry catalogEntry) (*Service, error) {
	directory, err := m.directoryFor(entry)
	if err != nil {
		return nil, err
	}
	service := m.newService(directory)
	if err := service.Init(); err != nil {
		return nil, err
	}
	id, found, err := service.WorkspaceID()
	if err != nil || !found || id != entry.ID {
		service.Shutdown()
		return nil, errors.New("workspace identity does not match its catalog entry")
	}
	return service, nil
}

func (m *WorkspaceManager) directoryFor(entry catalogEntry) (string, error) {
	if entry.RelativeDirectory == "." {
		return m.root, nil
	}
	parts := strings.Split(filepath.ToSlash(entry.RelativeDirectory), "/")
	if len(parts) != 2 || parts[0] != "workspaces" || parts[1] != entry.ID {
		return "", errors.New("workspace path is invalid")
	}
	if _, err := uuid.Parse(entry.ID); err != nil {
		return "", errors.New("workspace ID is invalid")
	}
	directory := filepath.Join(m.root, "workspaces", entry.ID)
	root := filepath.Clean(m.root) + string(os.PathSeparator)
	if !strings.HasPrefix(filepath.Clean(directory)+string(os.PathSeparator), root) {
		return "", errors.New("workspace path escapes the data directory")
	}
	return directory, nil
}
